pumpkin-solver/src/propagators/gcc/gcc_extended_resolution/inequality_sets.rs

```diff
@@ -0,0 +1,640 @@
+use crate::basic_types::Inconsistency;
+use crate::engine::propagation::LocalId;
+use crate::engine::propagation::PropagationContextMut;
+use crate::engine::propagation::Propagator;
+use crate::engine::propagation::ReadDomains;
+use crate::engine::DomainEvents;
+use crate::predicate;
+use crate::predicates::Predicate;
+use crate::predicates::PropositionalConjunction;
+use crate::variables::IntegerVariable;
+use crate::variables::Literal;
+
+pub(crate) struct GccInequalitySets<Var> {
+    variables: Box<[Var]>,
+    equalities: HashMap<(usize, usize), Literal>,
+}
+
+impl<Var> GccInequalitySets<Var> {
+    pub(crate) fn new(
+        variables: impl IntoIterator<Item = Var>,
+        equalities: HashMap<(usize, usize), Literal>,
+    ) -> Self {
+        Self {
+            variables: variables.into_iter().collect(),
+            equalities,
+        }
+    }
+
+    fn get_equality(&self, x: usize, y: usize) -> Literal {
+        assert!(x < self.variables.len());
+        assert!(y < self.variables.len());
+
+        *self
+            .equalities
+            .get(&(x, y))
+            .or(self.equalities.get(&(y, x)))
+            .expect("E_{x,y} or E_{y,x} must be defined")
+    }
+
+    fn get_inequality_explanation(&self, vars: &[usize]) -> Vec<Predicate> {
+        let mut reason = Vec::new();
+        for i in 0..vars.len() {
+            for j in (i + 1)..vars.len() {
+                let literal = self.get_equality(vars[i], vars[j]);
+                reason.push(!predicate!(literal == 0));
+            }
+        }
+        reason
+    }
+}
+
+impl<Var: IntegerVariable + 'static> Propagator for GccInequalitySets<Var> {
+    fn name(&self) -> &str {
+        "GCC extended resolution inequality sets"
+    }
+
+    fn initialise_at_root(
+        &mut self,
+        context: &mut crate::engine::propagation::PropagatorInitialisationContext,
+    ) -> Result<(), PropositionalConjunction> {
+        // In theory, we could run this propagator for every change in domains of variables
+        // but this would be rather wasteful, right?
+        for (i, var) in self.variables.iter().enumerate() {
+            let _ = context.register(var.clone(), DomainEvents::ASSIGN, LocalId::from(i as u32));
+        }
+
+        for ((i, j), literal) in self.equalities.iter() {
+            let id = (self.variables.len()) * (i + 1) + j;
+            // UPPER_BOUND changes -> the equality variable is assigned to 0
+            let _ = context.register(
+                *literal,
+                DomainEvents::UPPER_BOUND,
+                LocalId::from(id as u32),
+            );
+        }
+
+        Ok(())
+    }
+
+    fn debug_propagate_from_scratch(
+        &self,
+        mut context: PropagationContextMut,
+    ) -> crate::basic_types::PropagationStatusCP {
+        // Consider variables in random order
+        let mut variable_indices: Vec<usize> = (0..self.variables.len()).collect();
+        variable_indices.shuffle(&mut thread_rng());
```

Comment:
This introduces hard non-determinism, and makes our runs independent of seed.

Can we do some kind of heuristic sorting approach? My suggestion would be:
- Sort on number of inequalities related to each variable, break ties by domain size, then by initial variable order.

If you still think randomness is required, then at the very least we should use the existing randomness in the solver, so that it is still deterministic given a seed.